src/logics/SubstLoopBreaker.h

```diff
@@ -0,0 +1,192 @@
+    int       getLowLink()       const  { return lowlink; }
+    void      setStatus(NStatus s)      { status = s; }
+    NStatus   getStatus()        const  { return status; }
+    int       nChildren()        const  { if (children == TVLRef_Undef) return 0; return tvla[children].size(); }
+
+    SNRef&    operator[](int i)         { return tvla[children][i]; }
+    SNRef     operator[](int i)  const  { return tvla[children][i]; }
+    PTRef     getChildTerm(int i) const { return tvla[children].getChildTerm(i); }
+
+    SNRef     getNextChild();
+    void      updateLowLink(int i)      { lowlink = lowlink < i ? lowlink : i; }
+    bool      processed()         const { return tvla[children].processed(); }
+    bool      processing()        const { return tvla[children].processing(); }
+    void      setProcessed()            { tvla[children].setProcessed(); }
+    void      setProcessing()           { tvla[children].setProcessing(); }
+    bool      hasChildren()       const { return children != TVLRef_Undef; }
+    void      swipeChildren()            { children = TVLRef_Undef; }
+
+    void      clearTarjan()             { procChild = 0; index = -1; lowlink = -1; status = NStatus::unseen; }
+};
+
+class SubstNodeAllocator : RegionAllocator<uint32_t> {
+private:
+    static int substNode32Size() {
+        return sizeof(SubstNode) / sizeof(uint32_t);
+    }
+    TargetVarListAllocator& tla;
+    const Logic& logic;
+    Map<PTRef,TVLRef,PTRefHash> TargetToTargetVarListRef;
+    vec<PTRef> getVars(PTRef term) const;
+    Map<PTRef,SNRef,PTRefHash> SourceToSNRef;
+    vec<SNRef> SNRefs;
+public:
+    SubstNodeAllocator(TargetVarListAllocator& tla, const Logic& l, uint32_t start_cap) : RegionAllocator<uint32_t>(start_cap), tla(tla), logic(l) {}
+
+    SNRef alloc(PTRef tr, PTRef target);
+
+          SubstNode& operator[](SNRef r)       { return (SubstNode&)RegionAllocator<uint32_t>::operator[](r.x); }
+    const SubstNode& operator[](SNRef r) const { return (SubstNode&)RegionAllocator<uint32_t>::operator[](r.x); }
+    SNRef getSNRefBySource(PTRef ptr)    const { return SourceToSNRef[ptr]; }
+    void  clearTarjan();
+private:
+    SubstNode* lea(SNRef r) { return (SubstNode*)RegionAllocator<uint32_t>::lea(r.x); }
+
+};
+
+
+class TarjanAlgorithm {
+    vec<SNRef> controlStack;
+    vec<SNRef> tarjanStack;
+    SubstNodeAllocator& sna;
+    int index;
+    void addNode(SNRef n);
+  public:
+    TarjanAlgorithm(SubstNodeAllocator &sna) : sna(sna), index(0) {}
+    ~TarjanAlgorithm() { sna.clearTarjan(); }
+    vec<vec<SNRef>> getLoops(SNRef startNode);
+};
+
+class SubstLoopBreaker {
+private:
+
+    const Logic& logic;
+    Map<PTRef,bool,PTRefHash> seen;
+    TargetVarListAllocator tvla;
+    SubstNodeAllocator sna;
+
+public:
+    SubstLoopBreaker(const Logic& l) : logic(l), tvla(1024), sna(tvla, logic, 1024) {}
+    Map<PTRef,PtAsgn,PTRefHash> operator() (const vec<Map<PTRef,PtAsgn,PTRefHash>::Pair*>&& substs);
```

Comment:
I don't think the parameter should be `const &&`, const rvalue does not really make sense, since you cannot move from it, if it is const.
The same is true for `constructSubstitutionGraph`
In the implementation I don't think you need to move this parameters anywhere, so I think this should be just const lvalue reference `const &`, and a bunch of `std::move`s can be removed.

Reply:
Hm yes... My idea was to have the map consumed by the function since the whole point of the process is to replace an old map with a new map.  But I could buy the argument that I'm needlessly making the implementation more specialised and it would be better to keep it general.  In any case, also if my initial idea would be good the argument shouldn't be declared const.

Reply:
OK, I get the original idea, but I don't think the current implementation is achieving that goal.
I don't know what would be the best way to do that. But I think that in any case `const &&` should be replaced, because it is confusing and the double move from the other comment needs to be addressed.

Reply:
This issue should now have been addressed.